### PR TITLE
Add fixed data

### PR DESCRIPTION
**Pull request by steeevin88**
add the ability to display fixed stop data on SCEta

for example, the ACE Train's stop times are the same every day (barring holidays). rather than wasting API calls to the 511 API, we can just read them from a JSON file + display them on SCEta

resovles #13

### CODE REVIEW

**Comment by steeevin88** [comment]
image

this is what it looks like right now (assume the time was before 3:35PM)

File: server/app.py [hunk 1]
```diff
@@ -33,6 +34,16 @@
...
 
+try:
+    fixed_stops = {}
+    with open(args.fixed_stops, "r") as stream:
+        try:
+            fixed_stops = json.load(stream).items()
+        except Exception:
+            logging.exception("unable to load JSON file")
+except Exception:
+    logging.debug("JSON file with fixed stops was not provided")
```
Comment (evanugarte) [inline server/app.py]: how about a single try/except like
```py
fixed_stops = {}
try:
    with open(args.fixed_stops, "r") as stream:
        fixed_stops = json.load(stream).items()
except FileNotFoundError:
    logging.debug("JSON file with fixed stops was not provided, skipping")
except Exception:
    logging.exception("unable to load JSON file")
```

File: server/fixed_stops.json [hunk 2]
```diff
@@ -0,0 +1,62 @@
+{
+  "San Jose": [
```
Comment (evanugarte) [inline server/fixed_stops.json]: we only need this field right? the transit page only displays predictions from diridon

File: server/fixed_stops.json [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    "2024-02-03T23:35:00Z",
+    "2024-02-03T00:35:00Z",
+    "2024-02-03T01:35:00Z",
+    "2024-02-03T02:38:00Z"
```
Comment (evanugarte) [inline server/fixed_stops.json]: also shouldnt we just store the times at which the train leaves? since storing the date we would never use

the entire json file could be these 6 lines
```json
[
    "23:35:00",
    "00:35:00",
    "01:35:00",
    "02:38:00"
]
```

File: server/app.py (hunk 1 above)
Comment (steeevin88) [reply server/app.py]: resolved

File: server/fixed_stops.json (hunk 2 above)
Comment (steeevin88) [reply server/fixed_stops.json]: this makes sense, for some reason i thought we just wanted to display all the ACE data (from every stop location)

resolved

File: server/fixed_stops.json (hunk 3 above)
Comment (steeevin88) [reply server/fixed_stops.json]: i did this because the times we get back from 511 is an entire date string, storing the times only definitely makes more sense though

i decided to store them in PST so it's less code, i can explain if you want but hopefully that's okay

resolved

**Comment by steeevin88** [comment]
i turned all times to PST, for example
```json
{
    "Rapid Diridon Station": [
        "16:58:00",
        "17:18:00",
        "17:38:00"
    ]
}
```
and this is what the ACE page looks like
image

**Comment by SilverizedVGC** [comment]
(please ignore) I use the ACE train regularly LMAO

File: server/modules/args.py [hunk 4]
```diff
@@ -4,6 +4,7 @@
 def get_args():
     parser = argparse.ArgumentParser()
     parser.add_argument("--config", help="yaml file", required=True)
+    parser.add_argument("--fixed_stops", help="json file with predetermined stop data")
```
Comment (evanugarte) [inline server/modules/args.py]: the variable will still be `args.fixed_stops`, argparse is nice to us

```suggestion
    parser.add_argument("--fixed-stops", help="json file with predetermined stop data")
```

File: server/app.py [hunk 5]
```diff
@@ -68,10 +78,12 @@ class Cache:
   stops: typing.List[Stop]
   updated_at: str
 
+pst = pytz.timezone('America/Los_Angeles') 
 PREDICTIONS_URL = 'https://api.511.org/transit/StopMonitoring'
 cache = Cache([], '')
 def update_cache():
     new_stops = []
+    cache.updated_at = datetime.datetime.now(datetime.timezone.utc).astimezone(pst).time()
```
Comment (evanugarte) [inline server/app.py]: can we keep everything in utc? i dont think we need pytz anymore then

File: server/app.py [hunk 6]
```diff
@@ -118,7 +135,8 @@ def get_stop_predictions(stop_ids, operator, stop_name):
         all_incoming_buses = data.get('ServiceDelivery', {}).get('StopMonitoringDelivery', {}).get('MonitoredStopVisit')
         for bus in all_incoming_buses:
             route_name = bus.get('MonitoredVehicleJourney', {}).get('LineRef')
-            expected_arrival = bus.get('MonitoredVehicleJourney', {}).get('MonitoredCall', {}).get('AimedArrivalTime')
+            if (expected_arrival := bus.get('MonitoredVehicleJourney', {}).get('MonitoredCall', {}).get('AimedArrivalTime')):
```
Comment (evanugarte) [inline server/app.py]: i think we can undo these changes? since we wanna keep everything in utc for simplicity.

the frontend can be the one who converts stuff into the users local timezone

File: server/fixed_stops.json [hunk 7]
```diff
@@ -0,0 +1,6 @@
+[
+  "15:35:00",
```
Comment (evanugarte) [inline server/fixed_stops.json]: how about we make these time in utc

File: server/app.py (hunk 6 above)
Comment (steeevin88) [reply server/app.py]: resolved

File: server/fixed_stops.json (hunk 7 above)
Comment (steeevin88) [reply server/fixed_stops.json]: resolved

File: server/app.py (hunk 5 above)
Comment (steeevin88) [reply server/app.py]: resolved

File: server/modules/args.py (hunk 4 above)
Comment (steeevin88) [reply server/modules/args.py]: resolved

File: server/app.py [hunk 8]
```diff
@@ -67,11 +76,12 @@ class Stop:
 class Cache:
   stops: typing.List[Stop]
   updated_at: str
-
+  
 PREDICTIONS_URL = 'https://api.511.org/transit/StopMonitoring'
 cache = Cache([], '')
 def update_cache():
     new_stops = []
+    cache.updated_at = datetime.datetime.now(datetime.timezone.utc)
```
Comment (evanugarte) [inline server/app.py]: how about instead we keep the `cache.updated_at` line where it is, that way if there is some error during updating, we dont accidentally say the cache was updated at a certain time when it wasnt


instead on  this line we can define a variable called `now`, and then assign `cache.updated_at` to `now` at the bottom

File: server/app.py [hunk 9]
```diff
@@ -84,8 +94,36 @@ def update_cache():
...
+        # next train as coming tomorrow in UTC and recalculate the diff
+        if diff.days < 0:
+            dt = dt.replace(day=dt.day + 1)
+            diff = dt - cache.updated_at
+        
+        # if the difference is under 120 minutes, add it
+        if diff.seconds // 60 < 120:
+            filtered_times.append(dt)
+
+    stop_info = Stop([], "ACE Rail", [Prediction("ACE Train", {'San Jose' : filtered_times})] if filtered_times else [])
```
Comment (evanugarte) [inline server/app.py]: can we put the predictions for ACE under the diridon station id?

Comment (steeevin88) [reply server/app.py]: resolved

File: server/app.py (hunk 8 above)
Comment (steeevin88) [reply server/app.py]: resolved

**Review by evanugarte — APPROVED** [review]
very clean, thanks
